### PR TITLE
Attempt to normalize Location paths before creating URIs

### LINKED ISSUE #37737: ErrorLogger converts relative paths to incorrect URIs
> Looking at it in the debugger, the problem is that [this code](https://github.com/dotnet/roslyn/blob/548db2b498394680f4384a5c59125037312f1a56/src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs#L143) in Roslyn creates a System.Uri for the path, and URIs have different path normalization rules than paths (intentionally—see dotnet/corefx#31524).

> That transforms the legal path

> `S:\VS\src\vset\Agile\TestPlatform\sdk\ObjectModel\cs\\..\DataCollectorSettings.cs`

> into

> `file:///S:/VS/src/vset/Agile/TestPlatform/sdk/ObjectModel/cs/DataCollectorSettings.cs`

> which is also legal, but doesn’t point to the same place.

### PR DESCRIPTION
**Pull request by agocke**
The URI path parsing is not the same as Windows path parsing and
sometimes produces incorrect results. However, a canonical Windows
path should produce an equivalent URI. This change uses GetFullPath
for absolute paths passed to the ErrorLogger to normalize the path
before passing it to URIs. For relative paths we try to normalize
the directory separators, but there's no equivalent transformation
that guarantees correctness.

Fixes #37737

### CODE REVIEW

**Comment by agocke** [comment]
@dotnet/roslyn-compiler for review

**Comment by agocke** [comment]
ping @dotnet/roslyn-compiler

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/SarifErrorLoggerTests.cs [hunk 1]
```diff
@@ -83,5 +83,49 @@ public void DescriptorIdCollisionImpl()
             string actual = Encoding.UTF8.GetString(stream.ToArray());
             Assert.Equal(ExpectedOutputForDescriptorIdCollision, actual);
         }
+
+        protected void PathToUriImpl(string formatString)
+        {
+            var isUnix = PathUtilities.IsUnixLikePlatform;
+            var paths = new[] {
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/SarifErrorLoggerTests.cs]: Should this be a `readonly static` to amortize the allocation?

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/SarifErrorLoggerTests.cs [hunk 2]
```diff
@@ -83,5 +83,49 @@ public void DescriptorIdCollisionImpl()
...
+                (@"A:\B\C\\..\D.cs", isUnix ? @"A:/B/C/D.cs" : "file:///A:/B/D.cs"),
+                (@"A\B\C\\..\D.cs", isUnix ? @"A%5CB%5CC%5C%5C..%5CD.cs" : @"A/B/D.cs")
+            };
+
+            foreach (var (inputPath, outputPath) in paths)
+            {
+                var stream = new MemoryStream();
+
+                using (var logger = CreateLogger(
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/SarifErrorLoggerTests.cs]: Consider: `using var` declaration

File: src/Compilers/CSharp/Test/CommandLine/SarifV1ErrorLoggerTests.cs [hunk 3]
```diff
@@ -1,7 +1,12 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
+using System.Globalization;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/CommandLine/SarifV1ErrorLoggerTests.cs]: Are these using necessary?

Comment (agocke) [reply src/Compilers/CSharp/Test/CommandLine/SarifV1ErrorLoggerTests.cs]: Nope, removing.

Comment (jaredpar) [reply src/Compilers/CSharp/Test/CommandLine/SarifV1ErrorLoggerTests.cs]: Are you removing or just normalizing the change ...

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/SarifErrorLoggerTests.cs [hunk 4]
```diff
@@ -83,5 +83,47 @@ public void DescriptorIdCollisionImpl()
...
+            foreach (var (inputPath, outputPath) in paths)
+            {
+                var stream = new MemoryStream();
+
+                using var logger = CreateLogger(
+                        stream,
+                        toolName: "",
+                        toolFileVersion: "",
+                        toolAssemblyVersion: Version.Parse("1.0.0"),
+                        CultureInfo.InvariantCulture)
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/SarifErrorLoggerTests.cs]:
```suggestion
                        CultureInfo.InvariantCulture);
```

**Pull request merged**